### ISSUE TITLE
FeatureRequest: Linkcall exit early without timeout

### ISSUE THREAD

**Issue by ZTHawk** [node-red/node-red#4970]
### Current Behavior

When a linkcall node has no exit then there is a timeout.

Sometimes you want to exit the current flow early. But this is not possible when inside a linkcall.

### Expected Behavior

Some way to exit a linkcall early without a timeout error.
Maybe add a second output for linkcall nodes?

### Example flow

First inject works as intented.
Second inject should end without an error.
```
[{"id":"6a78e4b9cfcc0219","type":"inject","z":"6a6e954c.882fb4","name":"Linkcall ok","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"Linkcall","payload":"","payloadType":"date","x":180,"y":1450,"wires":[["1a351a8d0263bc70"]]},{"id":"2f068bc0443a4f2d","type":"debug","z":"6a6e954c.882fb4","name":"debug 15","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","statusVal":"","statusType":"auto","x":580,"y":1450,"wires":[]},{"id":"1a351a8d0263bc70","type":"link call","z":"6a6e954c.882fb4","name":"Test Linkcall no exit","links":["b153f886fae765dc"],"linkType":"static","timeout":"5","x":410,"y":1450,"wires":[["2f068bc0443a4f2d"]]},{"id":"b153f886fae765dc","type":"link in","z":"6a6e954c.882fb4","name":"TestLinkCall","links":[],"x":405,"y":1510,"wires":[["0f126feed175f4c5"]]},{"id":"d4a5574da4fe50b2","type":"link out","z":"6a6e954c.882fb4","name":"link out 109","mode":"return","links":[],"x":605,"y":1510,"wires":[]},{"id":"0f126feed175f4c5","type":"switch","z":"6a6e954c.882fb4","name":"","property":"topic","propertyType":"msg","rules":[{"t":"eq","v":"Linkcall","vt":"str"},{"t":"else"}],"checkall":"true","repair":false,"outputs":2,"x":500,"y":1510,"wires":[["d4a5574da4fe50b2"],[]]},{"id":"3d7d0297c270e8dd","type":"inject","z":"6a6e954c.882fb4","name":"Linkcall exit early","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"Linkcall_2","payload":"","payloadType":"date","x":200,"y":1500,"wires":[["1a351a8d0263bc70"]]}]
```

### Environment

- Node-RED version: 4.0.3

**Comment by debadutta98** [comment]
Hi @knolleary, Hope you are doing well!

I plan to work on this feature by providing options in the UI to configure early exit behavior. Here's how it can be implemented:  

1. **Enable Early Exit**:  
   - A radio button to enable or disable the early exit feature.  

2. **Early Exit Configuration**:  
   - If the user enables early exit, additional fields will be displayed to configure the behavior:  
     - **Early Exit Timeout**: A numeric input in second to specify the timeout duration for early exit. This value must be less than the overall timeout of the `link call` node.  
     - **Exit Flow Type**: A dropdown menu to choose the type of exit flow:  
       - **Static Exit Flow**: The user can select from predefined exit flows (e.g., specific node IDs or labels).  
       - **Dynamic Exit Flow**: The exit path will be dynamically determined at runtime based on the message payload. No further input is required from the user, as the path will be extracted directly from the payload.  

In the backend, I am considering implementing this feature using a race condition. If the original flow is executed and a response is received within the early exit timeout, the system will proceed with the original flow. Otherwise, it will follow the early exit flow

Please let me know if you have any suggestions or enhancements to include.

**Comment by knolleary** [comment]
@debadutta98 thanks for the interest. I think this feature needs more discussion before we start implementing anything.

It isn't clear from the original description what additional mechanism is desired - if not a timeout, then how is the exit triggered? If it is a timeout, then why isn't the existing timeout handling appropriate? Why have a second timeout when there is already a timeout that can be used.

From your proposal, the node should not be linking to other nodes from the UI. Any early exit should trigger a regular output of the node - not triggering other unconnected nodes.

(As it's a holiday, I may be slow to respond over the next few days)

**Comment by debadutta98** [comment]
@knolleary Agree with you, the original description does not clearly define the requirements or explain the issue in detail. So @ZTHawk can better elaborate further on their perspective.

Based on my understanding, I tried to reproduce the issue they described. I observed that if the link call node does not receive a response within the specified timeout, it throws a timeout error, which cause exit the flow.

To address this, I propose two potential solutions:

   1. Provide an option for users to configure a fallback flow that will execute if a timeout error occurs.
   2. Add another UI option to configure an early exit timeout. This timeout must be shorter than the original link call timeout. If the flow reaches the early exit timeout without receiving a response, the original flow would be aborted, and the system would proceed to the early exit flow instead.
 
    
![image](https://github.com/user-attachments/assets/2598e8b3-d308-4c2e-8e63-6d841911ce44)


Thank you for your response, @knolleary . Wishing you a Merry Christmas!

**Comment by ZTHawk** [comment]
Hello together,
My primary intention is to early exit a linkcall node without it throwing a timeout error and without it calling linked nodes.
So my example from first post (here a screenshot):
![grafik](https://github.com/user-attachments/assets/9870737f-8b44-488d-aa48-22bd4495229d)

So calling "Linkcall exit early" should neither throw an catable error, not call the debug node.
Alternativly add a second exit to linkcall node, which is triggered in case of early exit (which then can be ignored).

**Comment by debadutta98** [comment]
@ZTHawk As I understand it, are you suggesting adding a fallback option or an exit node for the link call? This would mean that if the link call throws an error, it would fall back to the exit node. 

Please correct me if I am wrong.

Thank you!

**Comment by knolleary** [comment]
The node already provides a timeout mode that can be handled with the Catch node - triggering a separate flow if the timeout occurs without any errors being logged. I don't immediately see the benefit of adding a different timeout.

**Comment by ZTHawk** [comment]
No not realy.

1. If the linknode throws an error then it should not be redirected to any additional exit. Just let it throw the error the normal way.
2. If the linkcall node does not return in time then throw a timeout
3. Exception from 2) If linkcall node runs into a dead-end (see screnshot the switch with its second exit) then no timeout should be thrown

Catching the timeout error right now is complecated because:

- you need to know if it is a regular exception or terminated early
- you need to add additional throw nodes into the linkcall flow (to identiy it) (if you have multiple early exists it could be messy)
- the catch node needs additional handling to identify if the error is a regular one or an early exit

So in general: I would like to use the funtionality to exit anywhere (without throwing errors; which is the current behaviour when in normal flow).
And additionally the feature to have reusable code (linkcall).

**Comment by knolleary** [comment]
A switch node after the catch node should be able to distinguish timeouts from any other errors. Timeouts are a type of error, so handling with the catch node is appropriate.

**Comment by ZTHawk** [comment]
Yes it it true, but sometimes a timeout is valid (which I do not want to filter) and sometimes it is not valid (my request) which would require filtering.

So whenever I want to exit a linknode early i need (right now):

1. Explicit throw node to specify that I am exiting early
2. Every catch node needs to handle those specific errors to prevent false positives

**Comment by knolleary** [comment]
How do you propose to trigger an early exit then? Sounds like it isn't just another timeout if you want timeouts to still occur and be handled as such.

**Comment by ZTHawk** [comment]
I have no real idea what is possible under the hood.

One possibility would be to have two exits, so I can connect the "early exit" code path to this new exit (this is probabely what @debadutta98 described with "exit flow type" -> "static exit flow").

**Comment by debadutta98** [comment]
To further clarify @ZTHawk 's point, if I understand correctly, the goal is to implement a use case like the following:

If we have two flows, Flow A and Flow B, and Flow A takes more time than expected, the system should switch to Flow B. @ZTHawk , please correct me if I’m mistaken.

We must set a threshold timeout and exit flow before executing the link call to achieve this. If Flow A exceeds the configured threshold timeout, the link call should terminate and move to Flow B.

CC: @knolleary 

Thank you!

**Comment by ZTHawk** [comment]
@debadutta98 this is not my request

Take a look at my example again. Flow A ("Linkcall ok") will call the debug node. If in theory the switch would take too long then for Flow A an error should be thrown. No fallback to anything.

Flow B ("Linkcall exit early") is running into a timeout. But from a flow logic point of view there is no reason for it. The execution should just stop without throwing a timeout (because we want the flow to stop after the switch). If Flow B would take too long to evaluate the switch then of course a timeout should be thrown.

Summary:
- no fallback
- allow timeout to be thrown (if legit)
- no timeout error if the linkcall flow exits (maybe by adding a second output)

**Comment by debadutta98** [comment]
@ZTHawk, I understand your point. Based on the flow you shared in the issue description, it seems you are suggesting the need for a node that functions as an end-of-flow node. Its main purpose would be to terminate the flow at any point and optionally return a result.

In your example, the second output of the switch node isn’t connected to anything, which is why it doesn’t return a result, leading to a timeout.

Approaches:

- We could consider introducing a new node specifically designed to terminate a flow at any point.
- Alternatively, an option could be added to the link out node to allow terminating the flow without returning a result.
- By default, if no node is attached in the else condition, the flow should terminate automatically.

The decision rests with @knolleary to determine which approach, or an alternative approach, should be followed to achieve the best outcome.

**Comment by ZTHawk** [comment]
@debadutta98 
Yes, now you got it 🙂
